Implement new platform plugin discovery.

azasypkin opened the pull request:
In this PR I want to introduce the skeleton for core plugin service + plugin discovery "module".

My current idea is to leverage existing `scanDirs` and `path` CLI/config arguments in the core for the convenience. In this case we'll have the following behavior (not implemented here yet, or maybe even I'll have a dedicated PR for that):

* `core` will be detecting its own plugins (the ones with `kibana.json`) and fail only if manifest exists, but it's invalid or if provided paths that are supposed to be directories don't exist or aren't directories. For all other cases (e.g. missing manifest) decision will be deferred to the legacy platform;

* legacy platform plugin discovery will work as before with one exception, it will consult with the `core` whether specific plugin is recognizable by the `core` and skip it if it's.

Part of https://github.com/elastic/kibana/issues/18840

azasypkin commented on src/core/server/plugins/plugin_discovery_error.ts:
```diff
@@ -0,0 +1,43 @@
+ * under the License.
+ */
+
+export enum PluginDiscoveryErrorType {
```
__note:__ I'm not entirely sure whether we need/want dedicated error type or not, but I found it handy to have one stream for all errors with the ability to filter out specific error type.

spalger replied:
I'm fine with it if it's helpful

azasypkin commented on src/core/server/plugins/plugins_discovery.ts:
```diff
@@ -0,0 +1,230 @@
+    return fsReadFile$(manifestPath).pipe(
+      mergeMap(manifestContent => {
+        try {
+          const plugin = { path, manifest: parseManifest(manifestContent) };
```
__note:__ I won't have `Plugin` class in this PR, but once it's implemented I'd probably be creating `Plugin` instances here instead of `{ path: string, manifest: PluginManifest }`.

azasypkin commented on src/core/server/plugins/plugins_discovery.ts:
```diff
@@ -0,0 +1,230 @@
+ */
+function parseManifest(rawManifest: Buffer): PluginManifest {
+  const manifest = JSON.parse(rawManifest.toString());
+  if (
```
__note:__ I made discovery a bit dumb so that it doesn't check plugin vs kibana versions match and defer it to the `PluginService`, it may change in the next PR, but what are your thoughts?

azasypkin replied:
Okay, noted!

azasypkin commented on src/core/server/plugins/plugins_service.ts:
```diff
@@ -0,0 +1,68 @@
+      this.configService.atPath('plugins', PluginsConfig).pipe(first())
+    );
+
+    await plugins$
```
__note:__ these two `await`'s just so that we actually trigger plugin discovery and make sure it's not broken accidentally by other changes, my plan is to have this in master, but I can get rid of them if you think it's too heavy/risky to have it in master until plugin service is fully functional.

azasypkin commented:
Hey @spalger @njd5475, PR is ready for the first feedback. Please share your thoughts, objections, ideas or any other concerns you may have :)

Thanks!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5496/)

azasypkin commented:
/cc @epixa

azasypkin commented on src/core/server/plugins/plugins_discovery.ts:
```diff
@@ -0,0 +1,230 @@
+   * all the errors that occurred during discovery process.
+   * @param config$ Plugin config instance.
+   */
+  public discover(config$: Observable<PluginsConfig>) {
```
Yeah, that's an interesting point, I'm not sure I can think of any use case though. But let's switch to a single config instance as an argument here.

azasypkin commented on src/core/server/plugins/plugins_service.ts:
```diff
@@ -0,0 +1,68 @@
+
+  constructor(
+    readonly env: Env,
+    readonly logger: LoggerFactory,
```
Good catch! Nope, I probably don't even need `env` here (at least at this point).

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5792/)

azasypkin commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5943/)

azasypkin commented:
Hey @spalger, I think I've handled the most of your other comments, except for these two https://github.com/elastic/kibana/pull/24256#discussion_r226917741 and  https://github.com/elastic/kibana/pull/24256#discussion_r226931740. Could you please take a look at my replies and let me know if you still prefer dedicated `discover` function instead of it being a class method and dedicated `package.json` existence check.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6587/)

spalger commented on src/core/server/plugins/plugins_discovery.ts:
```diff
@@ -0,0 +1,230 @@
+import { PluginDiscoveryError, PluginDiscoveryErrorType } from './plugin_discovery_error';
+import { PluginsConfig } from './plugins_config';
+
+const fsReadDir$ = bindNodeCallback(readdir);
```
Yeah, `bindNodeCallback()`, I don't know why it was so hard for me to wrap my head around but this cleared it up for me 🤓 [Used it](https://github.com/elastic/kibana/pull/24692/files#diff-a1b11423e8a99e5992dd9f589b7284d0R30) for the first time and it's so nice to have.

spalger commented on src/core/server/plugins/plugins_discovery.ts:
```diff
@@ -0,0 +1,230 @@
+              // Filter out non-directory entries from target directories, it's expected that
+              // these directories may contain files (e.g. `README.md` or `package.json`).
+              // We shouldn't silently ignore the entries we couldn't get stat for though.
+              mergeMap(pathStat => (pathStat.isDirectory() ? [path] : [])),
```
No, I don't think it is, and given the fact that canvas plugins have co-opted the plugins directory and they don't have package.json files maybe it's best that we don't worry about these errors.

spalger commented on src/core/server/plugins/plugins_discovery.ts:
```diff
@@ -0,0 +1,230 @@
+ * or/and direct paths to specific plugins.
+ */
+export class PluginsDiscovery {
+  constructor(private readonly log: Logger) {}
```
I generally prefer functions unless there is some tangible benefit of using a class, but I'm not picky. Using a class here doesn't cause issues, it just distracted me a bit while I wondered "why?" 😄

azasypkin replied:
Okay, since it's not a big deal for me and it already distracts you let's go with function for now.

azasypkin commented on src/core/server/plugins/plugin_discovery.test.ts:
```diff
@@ -0,0 +1,305 @@
+let discovery: PluginsDiscovery;
+beforeEach(() => {
+  mockReaddir.mockImplementation((path, cb) => {
+    if (path === '/scan/non-empty/') {
```
Yeah, they will likely fail, thanks. Will take a look if there's an easy solution

azasypkin replied:
Seems I found a solution that works for windows too. Not super simple, but these tests should not be updated that frequently. Alternative option was to have some custom jest snapshot serializer like we have in kbn-pm, but I don't really like this idea.

azasypkin commented on src/core/server/plugins/plugin_discovery_error.ts:
```diff
@@ -0,0 +1,43 @@
+  constructor(
+    public readonly type: PluginDiscoveryErrorType,
+    public readonly path: string,
+    public readonly cause?: Error
```
Yeah, looks like leftover

azasypkin commented on src/core/server/plugins/plugins_discovery.ts:
```diff
@@ -0,0 +1,241 @@
+    );
+
+    return {
+      plugins$: discoveryResults$.pipe(
```
Hmm, yeah, now when I know that `$` stands for `Stream` then it makes sense :slightly_smiling_face:

azasypkin commented on src/core/server/plugins/plugins_discovery.ts:
```diff
@@ -0,0 +1,241 @@
+    };
+  }
+
+  throw new Error('The "id" or/and "version" is missing in the plugin manifest.');
```
Yeah, it's a good idea, thanks!

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7528/)

azasypkin commented on src/core/server/plugins/discovery/plugin_discovery.test.ts:
```diff
@@ -0,0 +1,483 @@
+      "ui": false,
+      "version": "1",
+    },
+    "path": "${resolveForSnapshot(TEST_PATHS.scanDirs.nonEmpty, '1')}",
```
__note:__ I'm not super happy about this solution, but it works on windows too and we probably aren't going to change these tests a lot so :man_shrugging:

azasypkin replied:
> Do you know what jest does here if you use the -u flag?

I'm pretty sure, jest will just override without respecting my parameters.

> I think a basic .toEqual() would be better

Okay, I think you're right, will change it to something like this:
```ts
await expect(plugin$.pipe(toArray()).toPromise()).resolves.toEqual(
    [
      resolveForSnapshot(TEST_PATHS.scanDirs.nonEmpty, '1'),
      resolveForSnapshot(TEST_PATHS.scanDirs.nonEmpty, '3'),
      resolveForSnapshot(TEST_PATHS.scanDirs.nonEmpty2, '6'),
      resolveForSnapshot(TEST_PATHS.paths.existentDir),
      resolveForSnapshot(TEST_PATHS.paths.existentDir2),
    ].map(path => ({
      manifest: {
        id: 'plugin',
        kibanaVersion: '1.2.3',
        optionalPlugins: [],
        requiredPlugins: [],
        ui: false,
        version: '1',
      },
      path,
    }))
  );
```

azasypkin commented:
@spalger this one should be ready for another pass. I added version compatibility check and it was too much for a single file, so I moved discovery related files into `discovery` subfolder and extracted manifest parsing into separate file.

azasypkin commented on src/core/server/plugins/discovery/plugin_manifest_parser.ts:
```diff
@@ -0,0 +1,174 @@
+ * @param actualKibanaVersion Used Kibana version.
+ */
+function isVersionCompatible(expectedKibanaVersion: string, actualKibanaVersion: string) {
+  if (expectedKibanaVersion === ALWAYS_COMPATIBLE_VERSION) {
```
__note:__ I brought this from the legacy world, but not 100% sure we still want to keep that in the new platform, I'd say `yes`, but what do you think?

spalger replied:
👍 I think we should keep it for now

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7620/)

spalger commented on src/core/server/plugins/plugins_service.ts:
```diff
@@ -0,0 +1,69 @@
+  public async start() {
+    this.log.debug('starting plugins service');
+
+    const { error$, plugin$ } = await this.configService
```
Would you mind writing some integration tests to verify this works correctly? If you stub out `discover()` it should be pretty easy to test that errors, config and logging are all handled appropriately.

azasypkin replied:
Yep, sure

spalger commented on src/core/server/plugins/discovery/plugin_discovery.test.ts:
```diff
@@ -0,0 +1,483 @@
+`);
+});
+
+describe('parsing plugin manifest', () => {
```
Mind moving these to a test file for the plugin manifest parser?

azasypkin replied:
Sure, created tests before extracted parser into a separate file, but it makes sense to extract them as well now.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7643/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7813/)

azasypkin commented:
6.x/6.6: 9dbc08b32f918dac1e45994b1d357d7479a192a8
